### ISSUE TITLE
Scoped public packages not updating in NPM Enterprise

### ISSUE THREAD

**Issue by wesleytodd** [npm/npm#13163]
We have a [package on the public registry](https://www.npmjs.com/package/@streammedev/flyout) which is at `0.0.3` and our npm enterprise install still thinks it is `0.0.1`.  We specify `0.0.3` in our package json and the install fails when pointed to our internal registry but works when we go directly to the public registry.

UPDATE: Doing more debugging, it is actually difficult to find resources on how to debug issues like this.  Doesn't help that the `npme` package cannot remove it from our internal just so we can test with a fresh slate.  But my guess that is another issue...

**Comment by wesleytodd** [comment]
Also, when I try to manually remove the packages from the command line it fails:

```
$ npme remove-package "@streammedev/flyout"
Error dialing unix socket: dial unix /var/run/replicated.sock: no such file or directory
Error: dial unix /var/run/replicated.sock: no such file or directory
```

**Comment by wesleytodd** [comment]
For reference, after doing the update to make sure it wasnt something on our install, like a weird state, this is what I get when I try to remove the package:

```
$ npme remove-package @streammedev/flyout
Error dialing unix socket: dial unix /var/run/replicated.sock: no such file or directory
Error: dial unix /var/run/replicated.sock: no such file or directory
$ sudo npme remove-package @streammedev/flyout
Error reading ~/.replicatedcfg: read /root/.replicatedcfg: is a directory
Error: read /root/.replicatedcfg: is a directory
```

**Comment by othiym23** [comment]
Hey, @wesleytodd. While we'd like to make sure that your question gets answered, the CLI team doesn't really have a lot of expertise around npmE issues. I asked the npmE team, and they suggested taking this issue to [`npm/registry`](https://github.com/npm/registry/issues). Another option is to contact <mailto:support@npmjs.com> and point them at this issue for context. Either way, since this isn't a CLI issue, I'm going to close this issue, but please do follow up with us. Thanks for your time and your patience!

**Comment by wesleytodd** [comment]
Thanks for the direction, I had emailed support before and they had pointed me to the github repos.  Wasn't sure if it was a cli thing or a registry thing, so thanks for the help!!

**Comment by nexdrew** [comment]
@wesleytodd npm Enterprise now supports replication of public scoped packages (see https://github.com/npm/registry/issues/19#issuecomment-228116909) but it looks like you're running into a different problem related to the orchestration/scheduling layer used by npm Enterprise.

Would you mind running these commands and posting the output?

```
replicated --version
sudo ps aux | grep -e replicated
```

You may need to restart the Replicated service:

```
sudo service replicated restart || sudo systemctl restart replicated
```

**Comment by wesleytodd** [comment]
```
$ replicated --version
Replicated version 2.0.1626
```

Not super comfortable posting the entire output from the ps publicly, seems to have possibly sensitive info, but we have restarted replicated and replicated-ui multiple times.  It did have some errors which I can find if that would help, but a few retries and monitoring the logs was able to get it running without crashing.  There is some error that hasn't gone away and repeats a bunch: 

`Jun 23 17:30:24 redacted-hostname docker[6989]: ERROR 2016/06/23 17:30:24 authentication/authenticator.go:56 unexpected end of JSON input`

**Comment by wesleytodd** [comment]
Looking at the ps output, I am guessing this is what you need?

```
root      4855  0.0  0.1 178256  4536 ?        Ssl  Jun02   0:01 /usr/bin/docker run --name=replicated-operator ...<truncated>...
root      4895  0.2  0.3  58852 13964 ?        Ssl  Jun02  81:39 /usr/bin/replicated-operator
root      6989  0.0  0.6 178256 25216 ?        Ssl  17:24   0:00 /usr/bin/docker run --name=replicated  ...<truncated>...
root      7156  1.6  1.8 130824 73228 ?        Sl   17:24   0:30 /usr/bin/replicated -d
root      8477  0.0  0.6 178256 25436 ?        Ssl  17:26   0:00 /usr/bin/docker run --name=replicated-ui -p 8800:8800/tcp --volumes-from replicated quay.io/replicated/replicated-ui:latest
root      8519  0.2  0.5  33480 20324 ?        Ssl  17:26   0:03 /usr/bin/replicated-ui -d
```

**Comment by nexdrew** [comment]
@wesleytodd Thanks for that.

This sounds like an instance that started with Replicated 1.2 but was upgraded to Replicated 2.0. The difference is that Replicated 1.2 ran its services as native processes on the host while Replicated 2.0 runs its services as Docker containers, and the unix socket for Replicated is no longer used.

A side-effect of this change is that the `replicated` bin (which is used by the `npme` commands) moved from the host to a guest, and we "fix" this by using an alias on the host like this:

```bash
alias replicated="sudo docker exec -it replicated replicated"
```

Which is typically put into place by either the standard install process (on fresh install of `npme` >= 4) or via migration script:

```bash
curl -sSL https://get.replicated.com/migrate-v2 | sudo bash
```

The alias is typically kept in a `/etc/replicated.alias` file that is sourced via the global bashrc file (`/etc/bashrc` or `/etc/bash.bashrc`):

```bash
if [ -f /etc/replicated.alias ]; then
    . /etc/replicated.alias
fi
```

Perhaps your system ended up with a Replicated 1.2 bin even though you're obviously running Replicated 2.0. Could you verify with `which replicated`? I think you might want to run the migration script mentioned above, or at least replace the old `replicated` bin with the new alias.

Not sure how your system ended up in this state, but it's obviously something npm needs to handle better!

**Comment by wesleytodd** [comment]
Thanks so much for the help @nexdrew!  We do actually have the alias setup, as well as it being sourced properly.  I cant test the update script right now because this is our production registry and I will not break it for everyone just for this issue.  BUT, I can maybe try sometime late on a Friday where people wont be relying on it as much.

Is there a way to migrate the data from one install to another?  That way we could get into a fresh state and avoid whatever we did to get into this state?  If all we had to do was stand up a new instance and then migrate the data it would be much less nerve wracking for us since we wouldn't risk stopping the whole release and dev cycle.

**Comment by divolgin** [comment]
@wesleytodd do the contents of `/etc/replicated.alias` match what nexdrew posted?  Also check the output of the `alias` command in your shell and make sure it's also `sudo docker exec -it replicated replicated`.   And if the error has changed after everything you've done, please post the latest error output.

**Comment by wesleytodd** [comment]
Hey @divolgin, I did both of those but for reference here is what I just did:

```
$ cat /etc/replicated.alias
# THIS FILE IS GENERATED BY REPLICATED. DO NOT EDIT!

alias replicated="sudo docker exec -it replicated replicated"
$ alias replicated
alias replicated='sudo docker exec -it replicated replicated'
$ cat /etc/bash.bashrc | grep replicated
if [ -f /etc/replicated.alias ]; then
  . /etc/replicated.alias
$ sudo su
# alias replicated
alias replicated='sudo docker exec -it replicated replicated'
```

**Comment by nexdrew** [comment]
@wesleytodd Concerning this:

> Is there a way to migrate the data from one install to another? That way we could get into a fresh state and avoid whatever we did to get into this state? If all we had to do was stand up a new instance and then migrate the data it would be much less nerve wracking for us since we wouldn't risk stopping the whole release and dev cycle.

Good call! (And apologies for not warning about running the migration on a production instance!) You can start with a fresh install and either configure it to be a live replica of your primary instance or manually copy the data over. [See this unofficial documentation for more details.](https://gist.github.com/nexdrew/f00b16908743942196203b45abde70d8)

**Comment by nicholascapo** [comment]
I'm helping @wesleytodd with this issue.

Note the following:

```
$ alias replicated
alias replicated='sudo docker exec -it replicated replicated'

$ replicated --version
Replicated version 2.0.1626

$ which replicated
/usr/bin/replicated

$ /usr/bin/replicated --version
Replicated version 1.2.120
```

**Comment by divolgin** [comment]
@nicholascapo it appear you have 2 versions of replicated installed.  You can use the package manager to remove the old version.  And then re-run the curl script to fix up the service files, which will probably be removed by the package manager because they are in the same location between the two versions.

**Comment by nicholascapo** [comment]
The `replicated` alias will not be available in /usr/lib/node_modules/npme/lib/utils.js:17

```javascript
var exec = exports.exec = function (command, sudo, cb) {
  var commands = ['-c']
  if (sudo) command = 'sudo -E ' + command
  commands.push(command)

  var proc = spawn('sh', commands, {
    cwd: cwd,
    env: process.env,
    stdio: 'inherit'
  })

  proc.on('close', function (code) {
    cb(code)
  })
}
```

I replaced `/usr/bin/replicated` with the following shell wrapper:

```bash
#!/bin/bash
sudo docker exec -it replicated replicated $@
```

And then we were able to remove the package:

```bash
$ sudo npme remove-package @streammedev/flyout
removed package @streammedev/flyout
```

**Comment by nicholascapo** [comment]
@divolgin which deb packages can/should I remove?

```bash
$ dpkg -l | grep replicated
ii  replicated                     1.2-120                     amd64        Replicated
ii  replicated-agent               1.2-47                      amd64        Replicated Agent
ii  replicated-ui                  1.2-74                      amd64        Replicated web UI daemon
ii  replicated-updater             1.2-26                      amd64        Replicated Updater Daemon
```

**Comment by divolgin** [comment]
@nicholascapo all 4 are from old version.  Keep in mind this might remove service files too, and you will have to restore them with the curl script.

**Comment by wesleytodd** [comment]
Is there a reason the normal upgrade process did not do this correctly for
us? Until today we were not manually mucking with this machine.

**Comment by nexdrew** [comment]
@wesleytodd What was your method of upgrade? Just `sudo npm i npme -g --unsafe`?

**Comment by wesleytodd** [comment]
I don't remember using the unsafe flag. But otherwise yes. That and the ui
upgrade button. Also when we first started think it was called npmo. So we
did have to upgrade from that.

**Comment by nexdrew** [comment]
@wesleytodd There are a couple moving parts here:

1. Docker and Replicated

    These form the foundation of running npm Enterprise. They are typically installed or upgraded via our wrapper `sudo npm i npme -g --unsafe` but they can also be installed/upgraded individually/manually via their respective installation options.

    Note that running `sudo npm i npme -g` without the `--unsafe` flag will not work since `npm` will attempt to run the scripts as `nobody`, which does not have permissions to install/upgrade the necessary services.

2. The `npme` bin

    We use this as a convenience for (a) performing installs of the foundational services and (b) performing low-level admin tasks against a running instance.

    This is a normal npm package which is versioned/released separately from the npm Enterprise components themselves - it just has a special `postinstall` script to make the install process simpler.

    Once an instance is installed/running, this bin can be upgraded on its own (without affecting anything else) by running `sudo npm i npme -g --ignore-scripts`. This upgrades just the wrapper bin (skipping the `postinstall` script), which is occasionally needed if we add more convenience commands for managing features built into the npm Enterprise components.

3. The npm Enterprise components/services

    These are upgraded via the admin console on port 8800, distributed as versioned Docker images.

    Typically, this is the only thing you want to update regularly (every 2 weeks or so).

It just so happens that we recently updated our platform to move/upgrade the foundational services, which will rarely happen, and our story for smoothly making this transition is admittedly not great, as you have discovered. Internally, we have discussed ideas on how to better handle this, but our ideas have not been realized/implemented yet. I will use this issue as an opportunity to revisit our approach to foundational upgrades.

In the meantime, it's probably a good idea to setup a replica instance and schedule a maintenance window to perform the operations suggested by @divolgin. That will hopefully get things back to a normal state.

Thanks for reporting this! And please feel free to provide whatever feedback/suggestions you have.

**Comment by wesleytodd** [comment]
Thanks SO much for your help and the detailed response.  We were pretty much to understanding these moving parts from our poking around, but hearing a detailed description from you was helpful for sure.

We did run into the issue of running as `nobody`, but it was because npme told us to.  It was a nice colored message from `npme` that said something like, "an upgrade is available 4.0.0 -> 4.1.1, run npm install -g npme".  Don't know what repo to open an issue for that, but the message should probably include the flag.  I think we ended up clicking the button in the web ui and it just upgraded it for us, not sure I remember that correctly though.
